server/src/main/java/org/tctalent/server/service/db/impl/SavedSearchServiceImpl.java

```diff
@@ -931,768 +926,763 @@
-            savedSearch.setCountryIds(getListAsString(request.getCountryIds()));
-            savedSearch.setCountrySearchType(request.getCountrySearchType());
-            savedSearch.setSurveyTypeIds(getListAsString(request.getSurveyTypeIds()));
-            savedSearch.setEnglishMinSpokenLevel(request.getEnglishMinSpokenLevel());
-            savedSearch.setEnglishMinWrittenLevel(request.getEnglishMinWrittenLevel());
-            Optional<Language> language =
-                    request.getOtherLanguageId() != null ?
-                            languageRepository.findById(
-                                    request.getOtherLanguageId()) : Optional.empty();
-            if (language.isPresent()) {
-                savedSearch.setOtherLanguage(language.get());
-            }
-
-            Optional<SavedList> exclusionList =
-                    request.getExclusionListId() != null ?
-                            savedListRepository.findById(
-                                    request.getExclusionListId()) : Optional.empty();
-            if (exclusionList.isPresent()) {
-                savedSearch.setExclusionList(exclusionList.get());
-            }
-
-            savedSearch.setOtherMinSpokenLevel(request.getOtherMinSpokenLevel());
-            savedSearch.setOtherMinWrittenLevel(request.getOtherMinWrittenLevel());
-            savedSearch.setLastModifiedFrom(request.getLastModifiedFrom());
-            savedSearch.setLastModifiedTo(request.getLastModifiedTo());
-//        savedSearch.setCreatedFrom(request.getSearchCandidateRequest().getRegisteredFrom());
-//        savedSearch.setCreatedTo(request.getSearchCandidateRequest().getRegisteredTo());
-            savedSearch.setMinAge(request.getMinAge());
-            savedSearch.setMaxAge(request.getMaxAge());
-            savedSearch.setMinEducationLevel(request.getMinEducationLevel());
-            savedSearch.setEducationMajorIds(
-                    getListAsString(request.getEducationMajorIds()));
-            savedSearch.setMiniIntakeCompleted(request.getMiniIntakeCompleted());
-            savedSearch.setFullIntakeCompleted(request.getFullIntakeCompleted());
-
-            //Save Boolean filters corresponding to enum name
-            final CandidateFilterByOpps candidateFilterByOpps = request.getCandidateFilterByOpps();
-            if (candidateFilterByOpps == null) {
-                savedSearch.setAnyOpps(null);
-                savedSearch.setClosedOpps(null);
-                savedSearch.setRelocatedOpps(null);
-            } else {
-                savedSearch.setAnyOpps(candidateFilterByOpps.getAnyOpps());
-                savedSearch.setClosedOpps(candidateFilterByOpps.getClosedOpps());
-                savedSearch.setRelocatedOpps(candidateFilterByOpps.getRelocatedOpps());
-            }
+      //Construct and send emails
+      for (Long userId : userNotifications.keySet()) {
+        final Set<SavedSearch> savedSearches = userNotifications.get(userId);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.

Suggested remediation:
Access the values and keys with `Map.entry.getValue()` and `Map.entry.getKey()`, respectively, after iterating over the Map.entrySet().
```
@@ -1369,4 +1369,5 @@
       //Construct and send emails
-      for (Long userId : userNotifications.keySet()) {
-        final Set&lt;SavedSearch&gt; savedSearches = userNotifications.get(userId);
+      for (Map.Entry&lt;Long, Set&lt;SavedSearch&gt;&gt; entry : userNotifications.entrySet()) {
+        Long userId = entry.getKey();
+        final Set&lt;SavedSearch&gt; savedSearches = entry.getValue();
         String s = savedSearches.stream()

```